File: 07/Android/MyBleCentralApp/app/src/main/java/org/tamberg/myblecentralapp/MainActivity.java
// https://tamberg.mit-license.org/

package org.tamberg.myblecentralapp;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.BluetoothProfile;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanFilter;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Handler;
import android.os.Looper;
import android.os.ParcelUuid;
import android.provider.Settings;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import org.tamberg.util.Assert;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static android.bluetooth.BluetoothGatt.GATT_SUCCESS;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();
    private static final int REQUEST_ENABLE_BT = 1;
    private static final long SCAN_PERIOD_MS = 10000;
    // Bluetooth SIG registered 16-bit "UUIDs" have base UUID 0000xxxx-0000-1000-8000-00805f9b34fb
    private static final UUID HRM_SERVICE_UUID =
        UUID.fromString("0000180d-0000-1000-8000-00805f9b34fb");
    private static final UUID HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID =
        UUID.fromString("00002A37-0000-1000-8000-00805f9b34fb"); // N
    private static final UUID BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID =
        UUID.fromString("00002A38-0000-1000-8000-00805f9b34fb"); // R
    private static final UUID HEART_RATE_CONTROL_POINT_CHARACTERISTIC_UUID =
        UUID.fromString("00002A39-0000-1000-8000-00805f9b34fb"); // W

    private boolean mIsConnected = false;
    private BluetoothLeScanner mScanner;
    private BluetoothDevice mBluetoothDevice;
    private BluetoothGatt mBluetoothGatt;
    private Handler mHandler = new Handler(Looper.getMainLooper());

    private ScanCallback mScanCallback = new ScanCallback() {
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            super.onScanResult(callbackType, result);
            Log.d(TAG, "onScanResult, result = " + result.getDevice().getAddress());
            mBluetoothDevice = result.getDevice();
            connect(); // TODO: move to button handler
        }

        @Override
        public void onScanFailed(int errorCode) {
            Log.d(TAG, "onScanFailed, errorCode = " + errorCode);
        }
    };

    private final BluetoothGattCallback mGattCallback = new BluetoothGattCallback() {
        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Log.d(TAG, "onConnectionStateChange, STATE_CONNECTED\n\tgatt = " + gatt);
                mIsConnected = true;
                gatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Log.d(TAG, "onConnectionStateChange, STATE_DISCONNECTED\n\tgatt = " + gatt);
                disconnect();
            }
        }

        @Override
        public void onCharacteristicChanged(
            BluetoothGatt gatt, BluetoothGattCharacteristic characteristic)
        {
            Log.d(TAG, "onCharacteristicChanged, UUID = " +  characteristic.getUuid());
            UUID uuid = characteristic.getUuid();
            if (uuid.equals(HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID)) {
                //byte[] value = characteristic.getValue();
                int formatType = BluetoothGattCharacteristic.FORMAT_UINT8;
                int value = characteristic.getIntValue(formatType, 1);
                Log.d(TAG, "value = " + value);
            }
        }

        @Override
        public void onCharacteristicRead(
            final BluetoothGatt gatt, final BluetoothGattCharacteristic characteristic, int status)
        {
            Log.d(TAG, "onCharacteristicRead, UUID = " +
                characteristic.getUuid() + ", status = " + status);
            if (status == GATT_SUCCESS) {
                UUID uuid = characteristic.getUuid();
                if (uuid.equals(BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID)) {
                    byte[] value = characteristic.getValue();
                    Log.d(TAG, value.toString());
                }
            }
        }

        @Override
        public void onCharacteristicWrite(
            BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status)
        {
            Log.d(TAG, "onCharacteristicWrite, UUID = " +
                characteristic.getUuid() + ",\nstatus = " + status);
        }

        private void init (final BluetoothGatt gatt) {
            // TODO: implement a queue or use a 3rd party BLE library
            readCharacteristic(gatt, HRM_SERVICE_UUID,
                BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID, 100);
            writeCharacteristic(gatt, HRM_SERVICE_UUID,
                HEART_RATE_CONTROL_POINT_CHARACTERISTIC_UUID,
                0, BluetoothGattCharacteristic.FORMAT_UINT16, 500);
            setCharacteristicNotification(gatt, HRM_SERVICE_UUID,
                HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID, 1000);
        }

        @Override
        public void onServicesDiscovered(final BluetoothGatt gatt, int status) {
            Log.d(TAG, "Services discovered, status = " + status);
            if (status == GATT_SUCCESS) {
                init(gatt);
            }
        }
    };

    private void readCharacteristic(
        final BluetoothGatt gatt, final UUID serviceUuid, final UUID characteristicUuid, int delayMs)
    {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                BluetoothGattService gattService = gatt.getService(serviceUuid);
                if (gattService != null) {
                    BluetoothGattCharacteristic characteristic =
                        gattService.getCharacteristic(characteristicUuid);
                    if (characteristic != null) {
                        gatt.readCharacteristic(characteristic);
                    }
                }
            }
        }, delayMs);
    }

    private void writeCharacteristic(
        final BluetoothGatt gatt, final UUID serviceUuid, final UUID characteristicUuid,
        final int value, final int formatType, int delayMs)
    {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                BluetoothGattService gattService = gatt.getService(serviceUuid);
                if (gattService != null) {
                    BluetoothGattCharacteristic characteristic =
                        gattService.getCharacteristic(characteristicUuid);
                    if (characteristic != null) {
                        characteristic.setValue(value, formatType, 0);
                        gatt.writeCharacteristic(characteristic);
                    }
                }
            }
        }, delayMs);
    }

    private void setCharacteristicNotification(
        final BluetoothGatt gatt, final BluetoothGattCharacteristic characteristic, int delayMs)
    {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                gatt.setCharacteristicNotification(characteristic, true);
                // 0x2902 org.bluetooth.descriptor.gatt.client_characteristic_configuration.xml
                UUID configUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");
                BluetoothGattDescriptor descriptor = characteristic.getDescriptor(configUuid);
                descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
                gatt.writeDescriptor(descriptor);
            }
        }, delayMs);
    }

    private void setCharacteristicNotification(
        final BluetoothGatt gatt, UUID serviceUuid, UUID characteristicUuid, int delayMs)
    {
        setCharacteristicNotification(
            gatt, gatt.getService(serviceUuid).getCharacteristic(characteristicUuid), delayMs);
    }

    private boolean isLocationEnabled() {
        // Based on https://stackoverflow.com/questions/10311834
        boolean enabled;
        int locationMode = 0;
        try { // 19+
            locationMode = Settings.Secure.getInt(
                this.getContentResolver(), Settings.Secure.LOCATION_MODE);
            enabled = (locationMode != Settings.Secure.LOCATION_MODE_OFF);
        } catch (Settings.SettingNotFoundException e) {
            enabled = false;
        }
        return enabled;
    }

    private void scan() {
        Assert.check(mHandler != null);
        Assert.check(mScanner != null);
        List<ScanFilter> filters = new ArrayList<>();
        //filters.add(new ScanFilter.Builder().setDeviceAddress("C9:1E:3F:18:61:9D").build());
        filters.add(new ScanFilter.Builder().setServiceUuid(
            new ParcelUuid(HRM_SERVICE_UUID)).build()); // 21+
        ScanSettings settings = (new ScanSettings.Builder().setScanMode(
            ScanSettings.SCAN_MODE_LOW_LATENCY)).build();
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                Log.d(TAG, "stop scan");
                mScanner.stopScan(mScanCallback);
            }
        }, SCAN_PERIOD_MS);
        Log.d(TAG, "start scan");
        mScanner.startScan(filters, settings, mScanCallback);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d(TAG, "onCreate");
        boolean hasBle = getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE);
        // Or <uses-feature android:name="android.hardware.bluetooth_le" android:required="true"/>
        if (hasBle) {
            Log.d(TAG, "BLE available");
            BluetoothManager bluetoothManager =
                (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
            BluetoothAdapter bluetoothAdapter = bluetoothManager.getAdapter();
            if (bluetoothAdapter != null && bluetoothAdapter.isEnabled()) {
                Log.d(TAG, "BLE enabled");
                mScanner = bluetoothAdapter.getBluetoothLeScanner();
                String[] permissions = new String[]{
                    Manifest.permission.ACCESS_FINE_LOCATION
                };
                int requestCode = 0;
                ActivityCompat.requestPermissions(MainActivity.this, permissions, requestCode);
            } else {
                Log.d(TAG, "BLE not enabled");
                Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
                startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
            }
        } else {
            Log.d(TAG, "BLE not available");
        }
    }

    @Override
    public void onRequestPermissionsResult (
        int requestCode, String[] permissions, int[] grantResults)
    {
        if (isLocationEnabled()) {
            scan(); // TODO: move to button handler
        } else {
            Log.d(TAG, "Location not enabled");
        }
    }

    public void connect() {
        Log.d(TAG, "connect");
        Assert.check(mBluetoothDevice != null);
        Assert.check(mIsConnected == false);
        // call from main thread required to make connectGatt work
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                // work-around to prevent multiple calls to mGattCallback
                // based on http://stackoverflow.com/questions/33274009
                if (mBluetoothGatt == null) {
                    // see https://stackoverflow.com/questions/22214254
                    boolean autoConnect = false; // see onConnectionStateChange
                    Log.d(TAG, "mBluetoothDevice.connectGatt");
                    try {
                        mBluetoothGatt = mBluetoothDevice.connectGatt(
                            MainActivity.this, autoConnect, mGattCallback);
                        Log.d(TAG, "mBluetoothGatt = " + (mBluetoothGatt != null ?
                            mBluetoothGatt.toString() : "null"));
                    } catch (Exception e) {
                        Log.d(TAG, e.toString());
                    }
                } else {
                    Log.d(TAG, "mBluetoothGatt.connect");
                    mBluetoothGatt.connect();
                }
            }
        }, 1);
    }

    public void disconnect() {
        Log.d(TAG, "disconnect");
        mBluetoothGatt.disconnect();
        mBluetoothGatt.close();
        mBluetoothGatt = null;
        mIsConnected = false;
    }
}

File: 07/Android/MyBleCentralApp/app/src/main/java/org/tamberg/myblecentralapp/MainActivity.kt
// https://fhnw.mit-license.org/

package org.tamberg.myblecentralapp

import android.Manifest
import android.annotation.SuppressLint
import android.app.Application
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGatt.GATT_SUCCESS
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context.BLUETOOTH_SERVICE
import android.content.Context.LOCATION_SERVICE
import android.content.Intent
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.provider.Settings
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewmodel.compose.viewModel
import org.tamberg.myblecentralapp.ui.theme.MyBleCentralAppTheme
import java.util.*

val LocalBlePermissionHelper = staticCompositionLocalOf<MyBlePermissionHelper> { null!! }

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val helper = MyBlePermissionHelper(this) // hack
        setContent {
            MyBleCentralAppTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    CompositionLocalProvider(LocalBlePermissionHelper provides helper) {
                        MyBleCentralView()
                    }
                }
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    MyBleCentralAppTheme {
        MyBleCentralView()
    }
}

@Composable
fun MyBleCentralView(model: MyBleCentralViewModel = viewModel()) {
    val helper = LocalBlePermissionHelper.current // hack
    Column (
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center)
    {
        Row {
            Text(model.info.value)
        }
        Row {
            Button(
                enabled = model.enabled.value,
                onClick = { model.scan(helper) })
            {
                Text(model.command.value)
            }
        }
    }
}

class MyBlePermissionHelper(activity: ComponentActivity) {

    // helper class for functions that need a reference to the activity
    // there's probably a better way to do this right in Compose, e.g.
    // https://stackoverflow.com/questions/69075984

    lateinit var update: () -> Unit

    fun setUpdateCallback(updateCallback: () -> Unit) {
        update = updateCallback
    }

    private val startActivityForResultLauncher = activity.registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        update()
    }

    fun askToEnableBle() {
        val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
        startActivityForResultLauncher.launch(enableBtIntent)
    }

    fun askToEnableLocation() {
        val enableLocationIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
        startActivityForResultLauncher.launch(enableLocationIntent)
    }

    private val requestPermissionLauncher = activity.registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        update()
    }

    private val requestMultiplePermissionsLauncher = activity.registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        update()
    }

    // See https://developer.android.com/develop/connectivity/bluetooth/bt-permissions,
    // https://developer.android.com/reference/android/bluetooth/le/BluetoothLeScanner
    // #startScan(android.bluetooth.le.ScanCallback)
    fun askForPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            requestMultiplePermissionsLauncher.launch(arrayOf(
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_CONNECT))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            requestPermissionLauncher.launch(Manifest.permission.ACCESS_FINE_LOCATION)
        } else {
            requestPermissionLauncher.launch(Manifest.permission.ACCESS_COARSE_LOCATION)
        }
    }
}

class MyBleCentralViewModel(app: Application) : AndroidViewModel(app) {
    private val TAG = this.javaClass.name
    private val SCAN_PERIOD_MS: Long = 10000

    val info = mutableStateOf("Scan for peripherals.")
    val command = mutableStateOf("Scan")
    val enabled = mutableStateOf(true)

    // Note: This code has been converted from Java.

    private fun hasBle(): Boolean {
        val app = super.getApplication<Application>()
        return app.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)
    }

    private fun isBleEnabled(): Boolean {
        val app = super.getApplication<Application>()
        val bluetoothManager = app.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager
        val bluetoothAdapter = bluetoothManager.adapter
        return bluetoothAdapter != null && bluetoothAdapter.isEnabled
    }

    private fun isLocationEnabled(): Boolean {
        val result: Boolean
        val app = super.getApplication<Application>()
        val locationManager = app.getSystemService(LOCATION_SERVICE) as LocationManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            result = locationManager.isLocationEnabled
        } else {
            result = true
        }
        return result
    }

    // See https://developer.android.com/reference/android/bluetooth/le/BluetoothLeScanner#startScan(android.bluetooth.le.ScanCallback)
    private fun hasPermission(): Boolean {
        val result: Boolean
        val app = super.getApplication<Application>()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            result = (app.checkSelfPermission(Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED) &&
                    (app.checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED)
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            result = app.checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            result = app.checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
        } else {
            result = true // older versions had permission by default
        }
        return result
    }

    // Bluetooth SIG registered 16-bit "UUIDs" have base UUID 0000xxxx-0000-1000-8000-00805f9b34fb
    private val HRM_SERVICE_UUID =
        UUID.fromString("0000180d-0000-1000-8000-00805f9b34fb")
    private val HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID =
        UUID.fromString("00002A37-0000-1000-8000-00805f9b34fb") // N
    private val BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID =
        UUID.fromString("00002A38-0000-1000-8000-00805f9b34fb") // R
    private val HEART_RATE_CONTROL_POINT_CHARACTERISTIC_UUID =
        UUID.fromString("00002A39-0000-1000-8000-00805f9b34fb") // W

    private var mIsConnected = false
    private var mBluetoothDevice: BluetoothDevice? = null
    private var mBluetoothGatt: BluetoothGatt? = null

    @SuppressLint("MissingPermission")
    private fun readCharacteristic(
        gatt: BluetoothGatt, serviceUuid: UUID, characteristicUuid: UUID, delayMs: Long
    ) {
        val handler = Handler(Looper.getMainLooper())
        handler.postDelayed({
            val gattService = gatt.getService(serviceUuid)
            if (gattService != null) {
                val characteristic = gattService.getCharacteristic(characteristicUuid)
                if (characteristic != null) {
                    gatt.readCharacteristic(characteristic)
                }
            }
        }, delayMs)
    }

    @SuppressLint("MissingPermission")
    private fun writeCharacteristic(
        gatt: BluetoothGatt, serviceUuid: UUID, characteristicUuid: UUID,
        value: Int, formatType: Int, delayMs: Long
    ) {
        val handler = Handler(Looper.getMainLooper())
        handler.postDelayed({
            val gattService = gatt.getService(serviceUuid)
            if (gattService != null) {
                val characteristic = gattService.getCharacteristic(characteristicUuid)
                if (characteristic != null) {
                    characteristic.setValue(value, formatType, 0)
                    gatt.writeCharacteristic(characteristic)
                }
            }
        }, delayMs)
    }

    @SuppressLint("MissingPermission")
    private fun setCharacteristicNotification(
        gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, delayMs: Long
    ) {
        val handler = Handler(Looper.getMainLooper())
        handler.postDelayed({
            gatt.setCharacteristicNotification(characteristic, true)
            // 0x2902 org.bluetooth.descriptor.gatt.client_characteristic_configuration.xml
            val configUuid = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
            val descriptor = characteristic.getDescriptor(configUuid)
            descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
            gatt.writeDescriptor(descriptor)
        }, delayMs)
    }

    private fun setCharacteristicNotification(
        gatt: BluetoothGatt, serviceUuid: UUID, characteristicUuid: UUID, delayMs: Long
    ) {
        setCharacteristicNotification(
            gatt, gatt.getService(serviceUuid).getCharacteristic(characteristicUuid), delayMs
        )
    }

    private val mGattCallback: BluetoothGattCallback = object : BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Log.d(TAG, "onConnectionStateChange, STATE_CONNECTED\n\tgatt = $gatt")
                mIsConnected = true
                gatt.discoverServices()
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Log.d(TAG, "onConnectionStateChange, STATE_DISCONNECTED\n\tgatt = $gatt")
                disconnect()
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic
        ) {
            Log.d(TAG, "onCharacteristicChanged, UUID = " + characteristic.uuid)
            val uuid = characteristic.uuid
            if (uuid == HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID) {
                //byte[] value = characteristic.getValue();
                val formatType = BluetoothGattCharacteristic.FORMAT_UINT8
                val value = characteristic.getIntValue(formatType, 1)
                Log.d(TAG, "value = $value")
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onCharacteristicRead(
            gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int
        ) {
            Log.d(
                TAG, "onCharacteristicRead, UUID = " +
                        characteristic.uuid + ", status = " + status
            )
            if (status == GATT_SUCCESS) {
                val uuid = characteristic.uuid
                if (uuid == BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID) {
                    val value = characteristic.value
                    Log.d(TAG, value.toString())
                }
            }
        }

        override fun onCharacteristicWrite(
            gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int
        ) {
            Log.d(
                TAG, """
                onCharacteristicWrite, UUID = ${characteristic.uuid},
                status = $status
                """.trimIndent()
            )
        }

        private fun init(gatt: BluetoothGatt) {
            // TODO: implement a queue or use a 3rd party BLE library
            readCharacteristic(
                gatt, HRM_SERVICE_UUID,
                BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID, 100
            )
            writeCharacteristic(
                gatt, HRM_SERVICE_UUID,
                HEART_RATE_CONTROL_POINT_CHARACTERISTIC_UUID,
                0, BluetoothGattCharacteristic.FORMAT_UINT16, 500
            )
            setCharacteristicNotification(
                gatt, HRM_SERVICE_UUID,
                HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID, 1000
            )
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            Log.d(TAG, "Services discovered, status = $status")
            if (status == GATT_SUCCESS) {
                init(gatt)
            }
        }
    }

    @SuppressLint("MissingPermission")
    fun connect() {
        Log.d(TAG, "connect")
        assert(mBluetoothDevice != null)
        assert(mIsConnected == false)
        // call from main thread required to make connectGatt work
        val handler = Handler(Looper.getMainLooper())
        handler.postDelayed({
            // work-around to prevent multiple calls to mGattCallback
            // based on http://stackoverflow.com/questions/33274009
            if (mBluetoothGatt == null) {
                // see https://stackoverflow.com/questions/22214254
                val autoConnect = false // see onConnectionStateChange
                Log.d(TAG, "mBluetoothDevice.connectGatt")
                try {
                    val app = super.getApplication<Application>()
                    mBluetoothGatt = mBluetoothDevice!!.connectGatt(
                        app, autoConnect, mGattCallback
                    )
                    Log.d(
                        TAG,
                        "mBluetoothGatt = " + if (mBluetoothGatt != null) mBluetoothGatt.toString() else "null"
                    )
                } catch (e: Exception) {
                    Log.d(TAG, e.toString())
                }
            } else {
                Log.d(TAG, "mBluetoothGatt.connect")
                mBluetoothGatt!!.connect()
            }
        }, 1)
    }

    @SuppressLint("MissingPermission")
    fun disconnect() {
        Log.d(TAG, "disconnect")
        mBluetoothGatt!!.disconnect()
        mBluetoothGatt!!.close()
        mBluetoothGatt = null
        mIsConnected = false
    }

    private val scanCallback: ScanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            super.onScanResult(callbackType, result)
            Log.d(TAG, "onScanResult, result = " + result.device.address)
            mBluetoothDevice = result.device // pick first
            connect()
        }

        override fun onScanFailed(errorCode: Int) {
            Log.d(TAG, "onScanFailed, errorCode = $errorCode") // TODO
        }
    }

    @SuppressLint("MissingPermission")
    fun doScan() {
        val app = super.getApplication<Application>()
        val bluetoothManager = app.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager
        val bluetoothAdapter = bluetoothManager.adapter
        val scanner = bluetoothAdapter.bluetoothLeScanner
        val handler = Handler(Looper.getMainLooper())
        val filters: MutableList<ScanFilter> = ArrayList()
        //filters.add(new ScanFilter.Builder().setDeviceAddress("C9:1E:3F:18:61:9D").build());
        filters.add(ScanFilter.Builder().setServiceUuid(ParcelUuid(HRM_SERVICE_UUID)).build()) // 21+
        val settings = ScanSettings.Builder().setScanMode(
            ScanSettings.SCAN_MODE_LOW_LATENCY
        ).build()
        handler.postDelayed({
            Log.d(TAG, "stop scan")
            scanner.stopScan(scanCallback)
            enabled.value = true
        }, SCAN_PERIOD_MS)
        Log.d(TAG, "start scan")
        scanner.startScan(filters, settings, scanCallback)
    }

    private fun update() {
        if (!hasBle()) {
            info.value = "Bluetooth not available."
            command.value = "Scan"
            enabled.value = false
        } else if (!hasPermission()) {
            info.value = "Permission not given."
            command.value = "Give permission"
            enabled.value = true
        } else if (!isBleEnabled()) {
            info.value = "Bluetooth not enabled."
            command.value = "Enable Bluetooth"
            enabled.value = true
        } else if (!isLocationEnabled()) {
            info.value = "Location not enabled."
            command.value = "Enable location"
            enabled.value = true
        } else {
            info.value = "Scan for peripherals."
            command.value = "Scan"
            enabled.value = true
        }
    }

    fun scan(helper: MyBlePermissionHelper) {
        helper.setUpdateCallback { update() }
        enabled.value = false
        if (!hasBle()) {
            update()
        } else if (!hasPermission()) {
            helper.askForPermission()
        } else if (!isBleEnabled()) {
            helper.askToEnableBle()
        } else if (!isLocationEnabled()) {
            helper.askToEnableLocation()
        } else {
            doScan()
        }
    }
}
